### ISSUE TITLE
Producer.produceAsync performance

### ISSUE THREAD

**Issue by lukestephenson** [zio/zio-kafka#531]
I was making use of `produceAsync` in combination with `ZIO.foreach` to achieve batching / sending multiple records in parallel.

While I was expecting this to be slightly slower than the `produceChunkAsync` API, the order of magnitude of which it is slower was not.

For my experiment, I was trying to publish 1 million messages to Kafka (just a single broker / local docker). The timings I saw were:
| Experiment  | Duration |
| ------------- | ------------- |
| ZIO publishing using a chunk aware publisher (ie avoiding more ZIO invocations)  | 1.3 seconds  |
| ZIO publishing one message at a time, and relying on "sequence" to publish batches | 23 seconds  |
| Monix publishing of one message at a time, and relying on "sequence" to publish batches | 3.5 seconds |
| Fs2-kafka publishing one message at a time and relying on sequence. | 6 seconds |

For the monix, fs2-kafka implementations, I tried to keep it as close to ZIO-kafka as I could.

Note the fs2-kafka batch API is implemented in terms of the single publish API (and a `sequence`), so will be basically the same.

The monix implementation (our internal publisher lib, so I can't share, but I'll try to copy the relevant bits into the demo project soon) makes use of `scala.concurrent.blocking` to flag that calling the underlying `KafkaProducer.send` method may block.

One of my colleagues played with the zio-kafka implementation, and found that the majority of the performance hit is from [`ZIO.attemptBlocking`](https://github.com/zio/zio-kafka/blob/cc0c620da288aa87a5451d4a6a446532a1e6e2ff/zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala#L140).

For a properly configured KafkaProducer, the majority of calls to `send` shouldn't block, but the current implementation appears optimised for all the calls blocking, and always shifting execution onto another thread.

The code I used for the experiment is in https://github.com/lukestephenson/kafka-perf.

Now, you might be thinking just use the chunking API and the problem goes away.  Well yes to a certain extent.  But regardless of this I do have serious concerns about just how much overhead using ZIO imposes. Also, the current chunking API doesn't support what we need.  Our project has the ability to drop messages if publishing to kafka fails (eg message too large).  The zio-kafka bulk publish API doesn't tell you which message failed (just that something in the batch failed), or have the ability to continue with the other messages past that point.  So we would first need to improve that. Regardless of that, I don't think the `produceAsync` method should have overheads as high as it does currently.

**Comment by svroonland** [comment]
Thanks for your feedback @lukestephenson.

I'd say dropping the `attemptBlocking` should be safe, since `send` should not block like you mention. The `attemptBlocking` around most of the Consumer/Producer calls was likely added as a 'better safe than sorry', since it's hard to know if the implementation would block or not. Should it happen to block in some specific situation anyway, we can still rely on ZIO 2's automatic blocking detection.

**Comment by guizmaii** [comment]
@svroonland FYI, I was asking questions on this blocking subject here: https://github.com/zio/zio-kafka/issues/492

**Comment by svroonland** [comment]
@guizmaii Ah that `synchronized` is a good example of why it's better to be safe than sorry when interacting with possibly blocking Java libraries. ZIO 2's auto blocking detection changes things a bit, I think we can rely on that a bit more for performance critical pieces like here with the Producer.

**Comment by svroonland** [comment]
@adamgfraser Perhaps you could comment on the performance of `attemptBlocking`?

**Comment by svroonland** [comment]
@lukestephenson Ran your benchmark, below are results before and after removing attemptBlocking

Before:
```
Took chunked 1879
Took chunked 1326
Took chunked 865
Took chunked 930
Took chunked 895
Took chunked 925
Took unchunked 20784
Took unchunked 17760
Took unchunked 20199
Took unchunked 18230
Took unchunked 18247
Took unchunked 18151
```

After:
```
Took chunked 1755
Took chunked 1284
Took chunked 1055
Took chunked 1070
Took chunked 1011
Took chunked 1029
Took unchunked 4332
Took unchunked 3800
Took unchunked 3877
Took unchunked 3858
Took unchunked 3863
Took unchunked 4030
```

**Comment by lukestephenson** [comment]
Thanks for looking into this. Those results are similar to what we saw when removing `attemptBlocking`

**Comment by lukestephenson** [comment]
I took a further look into performance after switching to `attemptBlocking` as I was still finding that the zio-kafka performance was almost double what I would see with my Monix implementation.

It appears to be related to the performance of the promise implementation. Both the ZIO kafka and monix implementation rely on a Promise like contract (different implementations) to signal when the work has completed.
I've removed all the kafka logic and just timed how long the promises take to create.
For monix it looks like
```
import cats.effect.ExitCode
import monix.eval.{Task, TaskApp}

import scala.concurrent.Promise

object MonixPromisePerf extends TaskApp {
  override def run(args: List[String]): Task[ExitCode] = {
    val promiseLoop = Task.traverse((1 to 1_000_000).toList) { counter =>
      for {
        done <- Task.delay(Promise[Int]())
        _ = done.success(counter)
        _ <- Task.deferFuture(done.future)
      } yield done
    }

    val timed = promiseLoop.timed.flatMap(result => Task(println(s"Monix promiseLoop ${result._1.toMillis}")))
    val jobs = timed >> timed >> timed >> timed >> timed >> timed
    jobs.as(ExitCode.Success)
  }
}
```
For zio it looks like:
```
import zio.{ZIO, ZIOAppDefault, _}

object ZIOPromisePerf extends ZIOAppDefault {
  def run = {
    val promiseLoop = ZIO.foreach((1 to 1_000_000).toList) { counter =>
      for {
        done <- Promise.make[Throwable, Int]
        _ <- done.succeed(counter)
        _ <- done.await
      } yield done
    }

    val timed = promiseLoop.timed.flatMap(result => zio.Console.printLine(s"promiseLoop ${result._1.toMillis}"))
    timed.repeatN(5)
  }
}

```

In terms of performance, the Monix implementation averages 656ms, while the zio implementation takes 3100ms.  While the comparison is a bit naive / limited, it roughly correlates to the difference I see with my original timings between the kafka producing performance for monix / zio.

**Comment by guizmaii** [comment]
I think @adamgfraser would want to see these Promise performances comparaison

**Comment by guizmaii** [comment]
FYI https://github.com/zio/zio/pull/7557

**Comment by guizmaii** [comment]
@lukestephenson Could you maybe re-run your perfs tests with this version of ZIO: `2.0.4+19-064d1cd7-SNAPSHOT`, please?

To use a snapshot release you need to do this, in your `build.sbt`:
```scala
ThisBuild / resolvers ++= Resolver.sonatypeOssRepos("snapshots")

libraryDependencies += "dev.zio" %% "zio" % "2.0.4+19-064d1cd7-SNAPSHOT"
```

This is the version released with the Promise optimisation fix I mentioned previously (https://github.com/zio/zio-kafka/issues/531#issuecomment-1326535029)

**Comment by lukestephenson** [comment]
The promise performance on `2.0.4+19-064d1cd7-SNAPSHOT` is looking a lot better.  Locally it's down from 3100ms to 1100ms.

The kafka performance test has seen a similar reduction in time with the improvements to promise.  So with the improvements to promise and removing the use of `attemptBlocking`, kafka publishing time is down from 23 seconds to about 4 seconds.

**Comment by domartynov** [comment]
@svroonland I think the implemented solution in #555 is wrong. Is it a safe to drop blocking given `send` blocks periodically until published messages are committed (which is more like ~100ms - 1000ms in our prod for example). Should we add a perf warning to the API and suggest to use `produceChunkAsync`. With a volume causing the issue above, it probably does not make sense to use a promise per message anyway (compared to the cost of `Chunk` objects) which is confirmed by your test data where it's x4 slower even after dropping blocking.

**Comment by guizmaii** [comment]
What do you think about using a dedicated thread pool? 

Using the Blocking one isn't appropriate as it can be slow. 
Using the default one is inappropriate, as we might block its threads.
If we're using a dedicated one for zio-kafka, we can tune it and use a Thread pool optimised for it

WDYT?
🤔

**Comment by svroonland** [comment]
Ah that's really annoying, that methods that are explicitly documented to be asynchronous can still block, but you're right, it looks like there's [potential blocking while waiting for metadata](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L996). 

I believe the actual shifting from the current ZIO thread to another thread (from the blocking pool in this case) is causing the slowdown, not using the blocking thread pool per se. So using a dedicated thread pool won't help, I think.

Perhaps we can build your [earlier idea](https://github.com/zio/zio-kafka/pull/555#issuecomment-1324254280) of having a ZIO buffer (`Queue`) of `Chunk[ProducerRecord[Array[Byte], Array[Byte]]]`, combined with a dedicated fiber that is executed on the blocking pool which calls `send()` and completes a `Promise` for each `Chunk` of messages (which could be a Chunk of size 1 when using `produce`). That way we can be sure that calls to `produce` / `produceChunk` will never block and we don't have to shift to the blocking thread pool all the time.

**Comment by adamgfraser** [comment]
Why aren't we just running this whole thing on the blocking thread pool? Generally the biggest cost is in shifting between executors. Running a whole bunch of operations on the blocking executor and shifting there and back around the whole thing is dramatically faster than shifting there and back around every individual workflow.

**Comment by svroonland** [comment]
@adamgfraser If I understand you correclty, I think that's what I'm proposing; a send loop/stream running on the blocking  pool and a queue to communicate with the ZIO thread pool side. I hope to have some time to implement that later today.

**Comment by svroonland** [comment]
@guizmaii @adamgfraser https://github.com/zio/zio-kafka/pull/581

**Comment by guizmaii** [comment]
@lukestephenson Would it be possible for you to remove the dependency on `"com.zendesk" %% "gourmand"` in your https://github.com/lukestephenson/kafka-perf project maybe? 
It seems to be proprietary. I can't compile your code

**Comment by lukestephenson** [comment]
@guizmaii I'll remove the internal dependencies now

**Comment by guizmaii** [comment]
- 10 runs of the zio-kafka version with the latest version of dependencies:

<details>

```scala
sbt:kafka-perf> run;run;run;run;run;run;run;run;run;run
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1648ms
Took chunked 1142ms
Took chunked 760ms
Took chunked 772ms
Took chunked 767ms
Took chunked 752ms
Took unchunked 2457ms
Took unchunked 2457ms
Took unchunked 2228ms
Took unchunked 2297ms
Took unchunked 2342ms
Took unchunked 2593ms
[success] Total time: 22 s, completed 11 Jan 2023, 1:31:16 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1364ms
Took chunked 939ms
Took chunked 769ms
Took chunked 769ms
Took chunked 750ms
Took chunked 791ms
Took unchunked 2445ms
Took unchunked 2241ms
Took unchunked 2216ms
Took unchunked 2601ms
Took unchunked 2196ms
Took unchunked 2208ms
[success] Total time: 20 s, completed 11 Jan 2023, 1:31:36 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1653ms
Took chunked 988ms
Took chunked 739ms
Took chunked 845ms
Took chunked 817ms
Took chunked 817ms
Took unchunked 2935ms
Took unchunked 2911ms
Took unchunked 3089ms
Took unchunked 3234ms
Took unchunked 2763ms
Took unchunked 3080ms
[success] Total time: 24 s, completed 11 Jan 2023, 1:32:01 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1583ms
Took chunked 1329ms
Took chunked 1043ms
Took chunked 759ms
Took chunked 818ms
Took chunked 819ms
Took unchunked 2632ms
Took unchunked 2865ms
Took unchunked 4705ms
Took unchunked 3915ms
Took unchunked 2496ms
Took unchunked 4316ms
[success] Total time: 28 s, completed 11 Jan 2023, 1:32:28 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1971ms
Took chunked 1280ms
Took chunked 1069ms
Took chunked 1202ms
Took chunked 1062ms
Took chunked 1117ms
Took unchunked 4048ms
Took unchunked 3028ms
Took unchunked 2613ms
Took unchunked 2652ms
Took unchunked 3038ms
Took unchunked 3161ms
[success] Total time: 28 s, completed 11 Jan 2023, 1:32:56 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1557ms
Took chunked 941ms
Took chunked 770ms
Took chunked 805ms
Took chunked 732ms
Took chunked 801ms
Took unchunked 2438ms
Took unchunked 2432ms
Took unchunked 2594ms
Took unchunked 2254ms
Took unchunked 2322ms
Took unchunked 2329ms
[success] Total time: 21 s, completed 11 Jan 2023, 1:33:17 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1610ms
Took chunked 1017ms
Took chunked 766ms
Took chunked 933ms
Took chunked 728ms
Took chunked 801ms
Took unchunked 2464ms
Took unchunked 2294ms
Took unchunked 2197ms
Took unchunked 2150ms
Took unchunked 1929ms
Took unchunked 1903ms
[success] Total time: 19 s, completed 11 Jan 2023, 1:33:36 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1236ms
Took chunked 1050ms
Took chunked 738ms
Took chunked 1106ms
Took chunked 712ms
Took chunked 711ms
Took unchunked 2457ms
Took unchunked 2344ms
Took unchunked 2204ms
Took unchunked 2527ms
Took unchunked 2677ms
Took unchunked 2283ms
[success] Total time: 21 s, completed 11 Jan 2023, 1:33:57 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1558ms
Took chunked 1163ms
Took chunked 791ms
Took chunked 766ms
Took chunked 708ms
Took chunked 859ms
Took unchunked 2802ms
Took unchunked 2422ms
Took unchunked 2616ms
Took unchunked 2239ms
Took unchunked 2238ms
Took unchunked 2609ms
[success] Total time: 21 s, completed 11 Jan 2023, 1:34:18 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1545ms
Took chunked 1033ms
Took chunked 780ms
Took chunked 774ms
Took chunked 801ms
Took chunked 714ms
Took unchunked 2433ms
Took unchunked 2424ms
Took unchunked 2567ms
Took unchunked 2227ms
Took unchunked 2355ms
Took unchunked 2640ms
[success] Total time: 21 s, completed 11 Jan 2023, 1:34:39 pm
```

</details>

- 10 runs of the zio-kafka version with this version of the code: #582 (slower):

<details>

```scala
sbt:kafka-perf> run;run;run;run;run;run;run;run;run;run
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1920ms
Took chunked 1522ms
Took chunked 1280ms
Took chunked 1305ms
Took chunked 1290ms
Took chunked 1312ms
Took unchunked 4381ms
Took unchunked 4458ms
Took unchunked 3561ms
Took unchunked 4149ms
Took unchunked 3765ms
Took unchunked 3635ms
[success] Total time: 34 s, completed 11 Jan 2023, 1:46:08 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 2217ms
Took chunked 1558ms
Took chunked 1329ms
Took chunked 1249ms
Took chunked 1300ms
Took chunked 1274ms
Took unchunked 4363ms
Took unchunked 3810ms
Took unchunked 3792ms
Took unchunked 3889ms
Took unchunked 3683ms
Took unchunked 3659ms
[success] Total time: 33 s, completed 11 Jan 2023, 1:46:41 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1885ms
Took chunked 1668ms
Took chunked 1533ms
Took chunked 1331ms
Took chunked 1330ms
Took chunked 1347ms
Took unchunked 4370ms
Took unchunked 3964ms
Took unchunked 3974ms
Took unchunked 3769ms
Took unchunked 3787ms
Took unchunked 3834ms
[success] Total time: 33 s, completed 11 Jan 2023, 1:47:15 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 2204ms
Took chunked 1619ms
Took chunked 1462ms
Took chunked 1554ms
Took chunked 1456ms
Took chunked 1503ms
Took unchunked 4231ms
Took unchunked 4050ms
Took unchunked 3821ms
Took unchunked 3837ms
Took unchunked 3874ms
Took unchunked 3964ms
[success] Total time: 34 s, completed 11 Jan 2023, 1:47:49 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 2062ms
Took chunked 1492ms
Took chunked 1401ms
Took chunked 1386ms
Took chunked 1480ms
Took chunked 1379ms
Took unchunked 4308ms
Took unchunked 3832ms
Took unchunked 3773ms
Took unchunked 3656ms
Took unchunked 3902ms
Took unchunked 3651ms
[success] Total time: 33 s, completed 11 Jan 2023, 1:48:22 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1971ms
Took chunked 1414ms
Took chunked 1339ms
Took chunked 1370ms
Took chunked 1237ms
Took chunked 1390ms
Took unchunked 3984ms
Took unchunked 3881ms
Took unchunked 4124ms
Took unchunked 3940ms
Took unchunked 4021ms
Took unchunked 4046ms
[success] Total time: 33 s, completed 11 Jan 2023, 1:48:55 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1932ms
Took chunked 1350ms
Took chunked 1450ms
Took chunked 1269ms
Took chunked 1341ms
Took chunked 1270ms
Took unchunked 4278ms
Took unchunked 3717ms
Took unchunked 3933ms
Took unchunked 3691ms
Took unchunked 3737ms
Took unchunked 3924ms
[success] Total time: 32 s, completed 11 Jan 2023, 1:49:28 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 2410ms
Took chunked 1703ms
Took chunked 1360ms
Took chunked 1241ms
Took chunked 1423ms
Took chunked 1287ms
Took unchunked 4833ms
Took unchunked 3633ms
Took unchunked 3599ms
Took unchunked 3650ms
Took unchunked 3833ms
Took unchunked 3903ms
[success] Total time: 33 s, completed 11 Jan 2023, 1:50:01 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 2249ms
Took chunked 1563ms
Took chunked 1295ms
Took chunked 1440ms
Took chunked 1312ms
Took chunked 1431ms
Took unchunked 4965ms
Took unchunked 5195ms
Took unchunked 4161ms
Took unchunked 4413ms
Took unchunked 5060ms
Took unchunked 4920ms
[success] Total time: 39 s, completed 11 Jan 2023, 1:50:40 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 2483ms
Took chunked 2534ms
Took chunked 1311ms
Took chunked 1395ms
Took chunked 1324ms
Took chunked 1359ms
Took unchunked 4188ms
Took unchunked 3885ms
Took unchunked 4186ms
Took unchunked 4537ms
Took unchunked 3775ms
Took unchunked 3679ms
[success] Total time: 36 s, completed 11 Jan 2023, 1:51:15 pm
```

</details>

- 10 runs of the zio-kafka version with this version of the code https://github.com/zio/zio-kafka/pull/581 (same as initial/witness run):

<details>

```scala
sbt:kafka-perf> run;run;run;run;run;run;run;run;run;run
[info] compiling 1 Scala source to /Users/jules/conduktor/workspace/kafka-perf/target/scala-2.13/classes ...
[info] done compiling
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1360ms
Took chunked 1053ms
Took chunked 801ms
Took chunked 811ms
Took chunked 793ms
Took chunked 785ms
Took unchunked 3103ms
Took unchunked 2883ms
Took unchunked 2682ms
Took unchunked 2843ms
Took unchunked 2845ms
Took unchunked 3025ms
[success] Total time: 28 s, completed 11 Jan 2023, 1:53:24 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1269ms
Took chunked 1229ms
Took chunked 766ms
Took chunked 819ms
Took chunked 778ms
Took chunked 797ms
Took unchunked 2944ms
Took unchunked 2896ms
Took unchunked 2781ms
Took unchunked 2808ms
Took unchunked 2832ms
Took unchunked 2823ms
[success] Total time: 24 s, completed 11 Jan 2023, 1:53:48 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1328ms
Took chunked 1074ms
Took chunked 780ms
Took chunked 810ms
Took chunked 810ms
Took chunked 721ms
Took unchunked 3018ms
Took unchunked 2842ms
Took unchunked 2969ms
Took unchunked 3270ms
Took unchunked 2995ms
Took unchunked 2975ms
[success] Total time: 24 s, completed 11 Jan 2023, 1:54:12 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1417ms
Took chunked 957ms
Took chunked 817ms
Took chunked 797ms
Took chunked 810ms
Took chunked 813ms
Took unchunked 2938ms
Took unchunked 2822ms
Took unchunked 2577ms
Took unchunked 2806ms
Took unchunked 2821ms
Took unchunked 2622ms
[success] Total time: 23 s, completed 11 Jan 2023, 1:54:35 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1765ms
Took chunked 1069ms
Took chunked 769ms
Took chunked 839ms
Took chunked 846ms
Took chunked 833ms
Took unchunked 2975ms
Took unchunked 2548ms
Took unchunked 2804ms
Took unchunked 2799ms
Took unchunked 2549ms
Took unchunked 2853ms
[success] Total time: 23 s, completed 11 Jan 2023, 1:54:58 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1376ms
Took chunked 1008ms
Took chunked 787ms
Took chunked 797ms
Took chunked 805ms
Took chunked 701ms
Took unchunked 2986ms
Took unchunked 2535ms
Took unchunked 2689ms
Took unchunked 2697ms
Took unchunked 2707ms
Took unchunked 2741ms
[success] Total time: 23 s, completed 11 Jan 2023, 1:55:21 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1141ms
Took chunked 1086ms
Took chunked 749ms
Took chunked 771ms
Took chunked 707ms
Took chunked 777ms
Took unchunked 2803ms
Took unchunked 2733ms
Took unchunked 2522ms
Took unchunked 2719ms
Took unchunked 2786ms
Took unchunked 2499ms
[success] Total time: 22 s, completed 11 Jan 2023, 1:55:43 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1542ms
Took chunked 981ms
Took chunked 785ms
Took chunked 794ms
Took chunked 701ms
Took chunked 802ms
Took unchunked 3024ms
Took unchunked 2683ms
Took unchunked 2699ms
Took unchunked 2744ms
Took unchunked 2814ms
Took unchunked 2780ms
[success] Total time: 23 s, completed 11 Jan 2023, 1:56:06 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1215ms
Took chunked 1384ms
Took chunked 979ms
Took chunked 1040ms
Took chunked 905ms
Took chunked 921ms
Took unchunked 3326ms
Took unchunked 3214ms
Took unchunked 2978ms
Took unchunked 2679ms
Took unchunked 2869ms
Took unchunked 2906ms
[success] Total time: 25 s, completed 11 Jan 2023, 1:56:31 pm
[info] running ZIOKafkaPerf
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Took chunked 1538ms
Took chunked 933ms
Took chunked 890ms
Took chunked 821ms
Took chunked 784ms
Took chunked 786ms
Took unchunked 2901ms
Took unchunked 2560ms
Took unchunked 2772ms
Took unchunked 2766ms
Took unchunked 2556ms
Took unchunked 2775ms
[success] Total time: 23 s, completed 11 Jan 2023, 1:56:53 pm
```

</details>

**Comment by lukestephenson** [comment]
I've just pushed the kafka-perf.  Appologies for the delay.

**Comment by guizmaii** [comment]
Thanks Luke ❤️

**Comment by svroonland** [comment]
Shall we close this issue? Improvements were implemented in https://github.com/zio/zio/pull/7557, #555 and #581

**Comment by lukestephenson** [comment]
Firstly, thanks for all the fixes / effort investigated ❤️ 

I've updated the https://github.com/lukestephenson/kafka-perf repo to use the latest zio / zio-kafka dependencies and performance is much improved.  Note however, to see good performance in my case I had to bump the value of `ProducerSettings.sendBufferSize` (I think the default of 4096 is too low).

Here are the latest timings I collected (please don't compare to be original timings when I raised the issue, I'm running on a different laptop).

| Experiment  | Duration |
| ------------- | ------------- |
| ZIO publishing using a chunk aware publisher (ie avoiding more ZIO invocations)  | 1.3 seconds  |
| ZIO publishing one message at a time, and relying on "sequence" to publish batches | 3.4 seconds seconds  |
| Monix publishing of one message at a time, and relying on "sequence" to publish batches | 1.4 seconds |
| Fs2-kafka publishing one message at a time and relying on sequence. | Did not rerun |

With regards to the unchunked comparisons, originally ZIO was 6.6 times slower than my Monix implementation (23 vs 3.5 seconds). With the latest run, ZIO is only 2.4 times slower (using a sendBufferSize of 1mb - 3.4 vs 1.4 seconds).  Things have improved heaps, but it still appears to be a lot slower than the comparable Monix implementation. I'm happy for this issue to be closed. It's probably confusing to have the discussion continue any longer on this issue, so if we did want further improvements we can always raise another issue.  I suspect the slowness is because each message still needs to be pushed through a ZIO stream, and it isn't great with small chunk sizes.

Thanks again!

**Comment by guizmaii** [comment]
@lukestephenson Thanks again for all of this! 🙏 

I know that the ZIO team is re-implementing Streams on a new idea to improve performances (see Adam Fraser's talk at FunScala London 2022 (I don't know if it's published on YT yet)).

They told me that they wanted to deliver this new implementation in March/April for the next ZIO World event (https://www.eventbrite.com/e/zio-world-2023-lisbon-portugal-tickets-492267383997)
That'd be interesting to test the perfs of their new implementation when they release the first alpha/beta.

Meanwhile, I think the next step for us regarding performance will first be to implement a few benchmarks.
I don't mind keeping this issue open. I think it contains a good conversation we can continue to iterate on with future improvements.

**Comment by svroonland** [comment]
Renamed the issue because the previous title did not accurately reflect the latest performance comparison and can give false impressions to people not reading this entire thread.

**Comment by erikvanoosten** [comment]
@lukestephenson What are the latest performance numbers with ZIO 2.1.x?

I get the impression that we much closer to Monix' performance numbers now. But we'll probably never get there entirely since Monix used push-based streams and ZIO streams are pull based. I am beginning to think we should close this issue.

**Comment by erikvanoosten** [comment]
Since no one objected, I am closing this now.
